src/doc/TextDocument.ts

```diff
@@ -168,61 +167,87 @@ export default class TextDocument {
-    if (updated.length === affectedLines.length && updated.every((b, i) => b === affectedLines[i])) {
-      return this.selection === selection ? this : new TextDocument(this, selection);
+    function addLine(line: Line) {
+      line.length = line.content.length() + 1;
+      lines.push(line);
     }
 
-    lines = lines.concat(updated);
-    lines = lines.concat(lineIter.rest());
+    while (thisIter.hasNext() || otherIter.hasNext()) {
+      if (otherIter.peekType() === 'insert') {
+        const otherOp = otherIter.peek();
+        const index = typeof otherOp.insert === 'string' ? otherOp.insert.indexOf('\n', otherIter.offset) : -1;
+        if (index < 0) {
+          line.content.push(otherIter.next());
+        } else {
+          if (index) line.content.push(otherIter.next(index));
+          const newlineOp = otherIter.next(1);
+          const nextAttributes = line.attributes;
+          line.attributes = newlineOp.attributes || {};
+          addLine(line);
+          line = Line.create(undefined, nextAttributes, this.byId);
+        }
+      } else {
+        const length = Math.min(thisIter.peekLength(), otherIter.peekLength());
+        const thisOp = thisIter.next(length);
+        const otherOp = otherIter.next(length);
+        if (typeof thisOp.retain === 'number') {
+          if (throwOnError) throw new Error('apply() called with change that extends beyond document');
+          continue;
+        }
+
+        if (typeof otherOp.retain === 'number') {
+          const isLine = thisOp.insert === '\n';
+          let newOp: Op = thisOp;
+          // Preserve null when composing with a retain, otherwise remove it for inserts
+          const attributes = otherOp.attributes && AttributeMap.compose(thisOp.attributes, otherOp.attributes);
+          if (otherOp.attributes && !isEqual(attributes, thisOp.attributes)) {
+            if (isLine) {
+              line.attributes = attributes || {};
+            } else {
+              newOp = { insert: thisOp.insert };
+              if (attributes) newOp.attributes = attributes;
+            }
+          }
+          if (isLine) {
+            addLine(line);
+            line = Line.createFrom(thisIter.peekLine());
+          } else {
+            line.content.push(newOp);
+          }
+
+          // Optimization if at the end of other
+          if (otherOp.retain === Infinity || !otherIter.hasNext()) {
+            if (thisIter.opIterator.index !== 0 || thisIter.opIterator.offset !== 0) {
+              const ops = thisIter.restCurrentLine();
+              for (let i = 0; i < ops.length; i++) {
+                line.content.push(ops[i]);
+              }
+              addLine(line);
+              thisIter.nextLine();
+            }
+            lines.push(...thisIter.restLines());
+            break;
+          }
+        } // else ... otherOp should be a delete so we won't add the next thisOp insert
+      }
+    }
```

Comment:
Go through `op` by `op` between the change and the document and merge them together. The previous method was to take the affected lines and convert them into a Delta with the `id` attribute on the `\n' inserts and let the `Delta` class do the composition with `Delta.compose()` and convert it back into lines. This was easier and reused existing code, but it required `id` to be on `attributes` and didn't allow us to control how the ids were assigned to the lines. This method is more efficient since we skip the `Line` > `Delta` > `Line` conversion and more accurate since we can allow lines to keep their id, even if the `\n` insert changes to the next line.

It may be worth considering moving away from the `Delta` format and creating our own line-based data format. Or keeping Delta but not worrying about it being compatible with other editors that use it and put the newline _before_ the line it represents. So each document would start with a newline instead of end with one. Instead of:

```js
[
  { insert: 'This is my document\n' }
]
```

it would be 

```js
[
  { insert: '\nThis is my document' }
]
```

The benefit is that adding a newline in the middle of a paragraph will work semantically the way you expect. We wouldn't have to fight the format in our change operations.

Either of these ideas, however, would be a major breaking change. I know my data would break, and I'd have to do a huge data migration and somehow allow my system to work with both methods until the migration could roll out.

Reply:
I have no problems with changing how Typewriter consumes or interacts with Delta. My project is centered around raw text documents, so the delta format is simply an intermediate.